### DISCUSSION TITLE
Axios instance using accessToken from store

### DISCUSSION THREAD

**Discussion by miladdev85** [pmndrs/zustand#1582] [Q&A]
I'm struggling a bit with creating an Axios instance to use in my app. If accssToken is available, I want to use it from my Zustand Store. Is this the correct way of doing this?

```
// Use this if not authenticated yet.
const getClientAuth = () => {
  return encode(appConfig.basicHttpAuth.user + ':' + appConfig.basicHttpAuth.password);
};

// Axios instance will use Basic authentication as default. 
export const axiosInstance = axios.create({
  headers: { Authorization: `Basic ${getClientAuth()}` },
});

// Use Interceptor on request. Use accessToken from Zustand Store if there is one.
axiosInstance.interceptors.request.use(
  async config => {
    const { accessToken, apiRoot } = useAppStore.getState();

    if (config) {
      config.baseURL = apiRoot;
    }

    if (config?.headers) {
      if (accessToken) {
        config.headers.Authorization = `Bearer ${accessToken}`;
      }
      config.headers[APP_VERSION_HEADER] = APP_VERSION;
    }

    return config;
  },
  error => {
    return Promise.reject(error);
  },
);
```

**Comment by dai-shi** [comment]
I'm not very familiar with Axios API, but this looks good.
```js
    const { accessToken, apiRoot } = useAppStore.getState();
```

**Comment by VariabileAleatoria** [comment]
may I ask why would you store access token in zustand store?
I'm not sure wheter i would use it for piece of state that are not UI related (I never render them or use them to alter rendering)

**Comment by vincedsb1** [comment]
Hey @VariabileAleatoria,

It's not recommended to call `useStore` inside the `axiosInstance.interceptors.request.use` function. The reason is that hooks, like `useStore`, should only be called in React functional components or custom hooks, but not in regular JavaScript functions like interceptors. 

A common practice is to get the accessToken from the store once and pass it as a parameter to the interceptor function. Here's a modified version of your code:

```javascript
// Set up the interceptor with a parameter for accessToken
const requestInterceptor = (accessToken) => async (config) => {
  const { apiRoot } = useAppStore.getState();

  if (config) {
    config.baseURL = apiRoot;
  }

  if (config?.headers) {
    if (accessToken) {
      config.headers.Authorization = `Bearer ${accessToken}`;
    }
    config.headers[APP_VERSION_HEADER] = APP_VERSION;
  }

  return config;
};

// Create an axios instance
export const axiosInstance = axios.create({
  headers: { Authorization: `Basic ${getClientAuth()}` },
});

// Get the accessToken from Zustand Store
const { accessToken } = useAppStore.getState();

// Set up the interceptor with the accessToken
axiosInstance.interceptors.request.use(requestInterceptor(accessToken), error => {
  return Promise.reject(error);
});
```

By passing the `accessToken` as a parameter to the interceptor function, you ensure that the hook is only called in the appropriate context. This separation of concerns adheres to the rules of hooks in React.

Hope this helps! If it solves your issue, could you please mark this comment as the answer? It helps others find the solution faster. 🙏
